### ISSUE TITLE
Add support for statically linked binaries

### ISSUE THREAD

**Issue by Arinerron** [Arinerron/heaptrace#13]
Make a `heaptrace` program (not shared lib) with similar syntax to that of `ltrace` and add support for statically linked binaries.

* If the target binary is dynamically linked, use `LD_PRELOAD=./heaptrace.so` automatically as this is significantly more reliable/less hacky. Pass the argv before the target binary's name/args to `HEAPTRACE_ARGS` environ var.
* Otherwise, print an "unstable" warning. Then, resolve `malloc`/`free`/`realloc` symbols (require user to specify addresses if any syms aren't defined), use ptrace to set an `int3` breakpoint at those functions, and, when breakpoints get hit, call the `malloc`/`free`/`realloc` tracing functions.
  
  The `orig_malloc`/`orig_free`/`orig_realloc` functions in the static version of heaptrace should just resume execution until a `ret`, at which point it should get its return value (if relevant) and continue executing the tracing function. Once the tracing function is done, it should remove the relevant `ret` breakpoint and continue execution as if nothing ever happened.

**Comment by Arinerron** [comment]
tl;dr here's how to get statically linked binaries to work:

> basically just break at the start of malloc/free/realloc in the statically linked binary and call the tracer funcs. once the tracer func reaches the part where it needs the return value, it breaks at the original func's `ret`, continues exec until then, then switches back to the tracer func now that it knows the return val. once the tracer func is done, it removes the `ret` breakpoint and continues the program exec as if nothing ever happened

**Comment by Arinerron** [comment]
ok, for setting breakpoints in malloc: when you hit the breakpoint, write another trap insruction to immediately after the address stored at the top of the stack (the return address) to catch the eturn value.

**Comment by Arinerron** [comment]
well, get the ebp and the return address should be at a fixed offset from there. todo what about the stack canary

**Comment by Arinerron** [comment]
execve to start process, /proc/pid/maps to find binary base, parse elf structure to find symbol table, find symbols

**Comment by Arinerron** [comment]
done in ptrace branch
